[NEW] Added amdfftw, amdlibflame and amdscalapack recipes

vkallesh opened the pull request:
Updated base fftw, libflame and netlib-scalapack recipes
to accommodate the above listed AMD Optimizing CPU Libraries
which are a set of numerical routines optimized for AMD platforms.

Updated amdblis spack recipe

amdblis:
	1. updated with amdblis 2.2 release
amdfftw:
	1. "--enable-single" now work as synonym for "--enable-float"
amdlibflame:
	1. Added enable_or_disable_threads() to set value for "--enable-multithreading" flag
Libflame:
	1. Added enable_or_disable_threads() to set value for "--enable-multithreading" flag
	2. Corrected invocation of "enable_or_disable('threads')"

Change-Id: I9da0a2c2c4e2075b7fa2776e7cfe6548a2e0b32f

vkallesh committed:
[NEW] Added amdfftw, amdlibflame and amdscalapack recipes

Updated base fftw, libflame and netlib-scalapack recipes
to accommodate the above listed AMD Optimizing CPU Libraries
which are a set of numerical routines optimized for AMD platforms.

Updated amdblis spack recipe

amdblis:
	1. updated with amdblis 2.2 release
amdfftw:
	1. "--enable-single" now work as synonym for "--enable-float"
amdlibflame:
	1. Added enable_or_disable_threads() to set value for "--enable-multithreading" flag
Libflame:
	1. Added enable_or_disable_threads() to set value for "--enable-multithreading" flag
	2. Corrected invocation of "enable_or_disable('threads')"

Change-Id: I9da0a2c2c4e2075b7fa2776e7cfe6548a2e0b32f

vkallesh commented:
As AMD AOCC compiler is added, so I closed my old pull request - PR #18155 and raised this new pull request. 

Note:
amdfftw don't support pfft feature so base fftw was changed to accomidate this.

vkallesh committed:
Added amd-toolchain-support as maintainers

Added team github account amd-toolchain-support
as maintainers for all the recipes owned by
AMD Optimizing CPU Libraries (AOCL) team

Change-Id: I9a7969bd48fc42cfbb88dd7bd93e0802c6138582

adamjstewart commented on var/spack/repos/builtin/packages/amdblis/package.py:
```diff
@@ -17,8 +17,9 @@ class Amdblis(BlisBase):
+    maintainers = ['amd-toolchain-support']
+
     version('2.2', sha256='e1feb60ac919cf6d233c43c424f6a8a11eab2c62c2c6e3f2652c15ee9063c0c9')
-    version('2.1', sha256='3b1d611d46f0f13b3c0917e27012e0f789b23dbefdddcf877b20327552d72fb3')
```
Is there any reason not to keep the old version? We generally keep old versions in packages in case someone needs to reproduce an older installation.

vkallesh replied:
In Spack, AOCC recipe support starts from 2.2 release.

To align with AOCC spack recipe,  AMD Optimizing CPU Libraries (AOCL) team decides to support all their libraries starting from 2.2 version onwards.

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,125 @@
+            '--enable-threads'
+            ]
+
+        # Check if coimpiler is AOCC
```
```suggestion
        # Check if compiler is AOCC
```

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,125 @@
+    depends_on('texinfo')
+    provides('fftw-api@3', when='@2:')
+
+    conflicts('precision=quad', when='%aocc', msg='AOCC clang doesn\'t support quad precision')
```
```suggestion
    conflicts('precision=quad', when='%aocc', msg="AOCC clang doesn't support quad precision")
```

vkallesh replied:
Required changes done and pushed again.

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,125 @@
+
+    conflicts('precision=quad', when='%aocc', msg='AOCC clang doesn\'t support quad precision')
+    conflicts('+debug', when='%aocc', msg='AOCC clang doesn\'t support debug')
+    conflicts('+openmp', when='%apple-clang', msg="Apple's clang does not support OpenMP")
```
Technically, you can get it to work. I did this in `py-scikit-learn`. But I can totally understand if you don't want to support something hacky like that.

vkallesh replied:
Handled apple-clang support with llvm-openmp

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,125 @@
+
+        # Check if coimpiler is AOCC
+        if spec.satisfies('%aocc'):
+            options.append('CC=clang')
+            options.append('CXX=clang++')
+            options.append('FC=flang')
```
Spack already sets `CC` and friends to the compiler wrappers, is this actually needed?

vkallesh replied:
Without these options amdfftw faced issues during compilation, so It is required.

vkallesh replied:
Updated the aocc clang selection as per spack standards

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,125 @@
+
+        if '+mpi' in spec:
+            options.append('--enable-mpi')
+            options.append('--enable-amd-mpifft')
```
I would explicitly disable all of these in else clauses

vkallesh replied:
Added suggested else clause

adamjstewart commented on var/spack/repos/builtin/packages/amdlibflame/package.py:
```diff
@@ -0,0 +1,66 @@
+
+    provides('flame@5.2', when='@2:')
+
+    depends_on('blas')
```
The base class already has a dependency on blas, so this line isn't needed

vkallesh replied:
Removed the line depends_on('blas')

adamjstewart commented on var/spack/repos/builtin/packages/amdlibflame/package.py:
```diff
@@ -0,0 +1,66 @@
+        args.append("--enable-external-lapack-interfaces")
+        return args
+
+    def build(self, spec, prefix):
+        """make function"""
+        make()
```
This is the default so this can be removed

vkallesh replied:
Removed

adamjstewart commented on var/spack/repos/builtin/packages/stream/package.py:
```diff
@@ -1,42 +0,0 @@
-from spack import *
-
-
-class Stream(MakefilePackage):
```
Was this package removed by accident or on purpose?

vkallesh replied:
Stream test is still on, so I deleted stream recipe purposefully

adamjstewart replied:
I'm not sure I understand, why do we want to remove this package for all users?

vkallesh replied:
My bad, I thought local updation of stream was removed.

I have now readded the stream again.

vkallesh committed:
Incorporated review comments

Updated packages.yaml with aocl components
Handled Flake8 test failures

Change-Id: I0a03f02d8c9f326b2434ec907958c3de3a8e18eb

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,133 @@
+            options.append('CXX=clang++')
+            options.append('FC=flang')
+
+        if '+shared' in spec:
+            options.append('--enable-shared')
+
+        if '+openmp' in spec:
+            options.append('--enable-openmp')
```
Can these also get else clauses? Explicit > implicit.

vkallesh replied:
Updated amdfftw with else clauses.

adamjstewart commented on var/spack/repos/builtin/packages/fftw/package.py:
```diff
@@ -126,6 +96,11 @@ def configure(self, spec, prefix):
 
         # Variants that affect every precision
         if '+openmp' in spec:
+            # Note: Apple's Clang does not support OpenMP.
+            if spec.satisfies('%clang'):
+                ver = str(self.compiler.version)
+                if ver.endswith('-apple'):
+                    raise InstallError("Apple's clang does not support OpenMP")
```
This can be converted to a conflict for `%apple-clang`, the compiler is no longer called `%clang`

vkallesh replied:
Currently apple-clang section is redundant, already it is handled in the conflict checks.
So removed that section under base fftw recipe.

vkallesh committed:
Readded accidental removal of stream recipe

amdfftw:
	1. Updated the aocc clang selection as per spack standards
fftw:
	1. Currently apple-clang section is redundant,
	already it is handled in the conflict checks.

Change-Id: Idef4a3f61717eb81f321e0cd16e7ba9619eac846

alalazo commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,144 @@
+
+    version('2.2', sha256='de9d777236fb290c335860b458131678f75aa0799c641490c644c843f0e246f8')
+
+    variant(
+        'precision', values=any_combination_of(
+            'float', 'single', 'double', 'long_double',
+            'quad').prohibit_empty_set().with_default('float,double'),
+        description='Build the selected floating-point precision libraries'
+    )
```
Is this value the same as in the base class? If so it shouldn't be needed to repeat it.

vkallesh replied:
We have additional value 'single' than base fftw. So variant section required.

alalazo replied:
Looking at the recipe it seems `single` is the same as `float`. Did I miss something?

vkallesh replied:
Yes, single and float are same. We are trying to provide more flexibility during usage.

alalazo replied:
If they are the same thing, please remove this variant entirely. Having `single` doing the same as `float` will change the hash of the spec (which may result in duplicated builds) but won't do anything to the underlying artifacts.

vkallesh replied:
Removed 'single' variant from amdfftw recipe

vkallesh committed:
Fix for style and docs/validate (pull_request) test

unnumbered format placeholders from {} to {0}

Change-Id: If67a3374177ec067573e5504462d257712fafc05

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,144 @@
+        # Check if compiler is AOCC
+        if spec.satisfies('%aocc'):
+            options.append(
+                "CC={0}".format(os.path.basename(self.compiler.cc))
```
`self.compiler.cc` is the actual compiler, not the compiler wrapper. Is it possible to get it working with Spack's compiler wrapper (`spack_cc`) instead?

vkallesh replied:
Changed the compiler reference to spack's compiler wrappers

vkallesh committed:
changed compiler references to Spack's compiler wrapper:spack_cc, spack_cxx, spack_fc

Change-Id: I7ae29c978fff16e37773913f14c84df232499763

alalazo reviewed (changes requested):
An initial request on the precision variant

vkallesh committed:
Removed 'single' variant from amdfftw recipe

Instead of conflict for apple-clang + openmp, handled this senario
via below available feature:
depends_on('llvm-openmp', when='%apple-clang +openmp')

Change-Id: I701b23d83e822a500ca3aaf2b60cc9ace09e13dc

vkallesh committed:
Added relevant info for users who prefers to use single precision

Change-Id: I3506e21da428ddef5fb7895b5aaed32c2a061ef6

vkallesh commented:
> An initial request on the precision variant

Changes are implemented, please review

vkallesh commented:
Closed by mistake, reopened again.
Note: all review comments are addressed.

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,130 @@
+    AMD Optimized FFTW is the optimized FFTW implementation targeted
+    for AMD CPUs.
+
+    For \'single\' precision build, please use precision value as float.
+    Example : \'spack install amdfftw precision=float\'
```
There's no need to escape the single quotes here.

vkallesh replied:
Remove escape sequence slash

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,130 @@
+
+    provides('fftw-api@3', when='@2:')
+
+    conflicts('precision=quad', when='%aocc', msg="AOCC clang doesn\'t support quad precision")
+    conflicts('+debug', when='%aocc', msg="AOCC clang doesn\'t support debug")
```
Same here

vkallesh replied:
Remove escape sequence slash

adamjstewart commented on var/spack/repos/builtin/packages/amdfftw/package.py:
```diff
@@ -0,0 +1,130 @@
+
+    conflicts('precision=quad', when='%aocc', msg="AOCC clang doesn\'t support quad precision")
+    conflicts('+debug', when='%aocc', msg="AOCC clang doesn\'t support debug")
+    conflicts('%gcc@:7.2', when="@2.2:", msg="Recommended GCC version above 7.2 for AMDFFTW")
```
Is this a recommendation or a requirement? `conflicts` is a hard requirement. If it's only a recommendation, I would remove this line.

vkallesh replied:
Rewording done from 'Recommended' to 'Required'

adamjstewart commented on var/spack/repos/builtin/packages/fftw/package.py:
```diff
@@ -202,3 +172,39 @@ def check(self):
+
+    variant(
+        'pfft_patches', default=False,
+        description='Add extra transpose functions for PFFT compatibility')
```
I usually like to put things in the following order:

1. versions
2. variants
3. dependencies
4. providers
5. patches

Providers aren't very common, so their dependency varies a bit more, but the rest are usually in this order. This isn't a requirement, just a suggestion for consistency with other packages.

vkallesh replied:
Reording done

adamjstewart commented on var/spack/repos/builtin/packages/libflame/package.py:
```diff
@@ -119,3 +106,23 @@ def darwin_fix(self):
+    version('5.2.0', sha256='997c860f351a5c7aaed8deec00f502167599288fd0559c92d5bfd77d0b4d475c')
+    version('5.1.0', sha256='e7189b750890bd781fe773f366b374518dd1d89a6513d3d6261bf549826384d1')
+
+    provides('flame@5.2', when='@5.2.0')
```
What about when at 5.1.0? I assume it still provides flame.

vkallesh replied:
Added provides entry for 5.1.0 version

adamjstewart reviewed (approved):
Some minor formatting suggestions, otherwise looks good

vkallesh committed:
Minor changes on fftw, amdfftw and libflame

amdfftw:
	1. Removed escape symbol to the single quotes
	2. Rewording the conflict line from Recommended
	to Required

fftw:
	1. Reorded to following recommended sections:
	versions, variants, dependencies, providers,
	patches

libflame:
	1. Added provides entry for 5.1.0 version

Change-Id: I21ebff99b6dfde031763154693ecb3f1fa47b476

vkallesh committed:
Removed single quote from amdfftw docstring to fix style failures

Change-Id: Ife939a5a2f5ccbc8879b730c7bebfe2fcfef9332

adamjstewart commented:
Can you resolve the merge conflicts?

vkallesh committed:
Merge branch 'develop' into amd_aocl_recipes
